// @vkontakte/vkui/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function get() {
    return _Root.default;
  }
});
Object.defineProperty(exports, "View", {
  enumerable: true,
  get: function get() {
    return _View.default;
  }
});
Object.defineProperty(exports, "Panel", {
  enumerable: true,
  get: function get() {
    return _Panel.default;
  }
});
Object.defineProperty(exports, "HeaderButton", {
  enumerable: true,
  get: function get() {
    return _HeaderButton.default;
  }
});
Object.defineProperty(exports, "PanelHeader", {
  enumerable: true,
  get: function get() {
    return _PanelHeader.default;
  }
});
Object.defineProperty(exports, "PanelHeaderContent", {
  enumerable: true,
  get: function get() {
    return _PanelHeaderContent.default;
  }
});
Object.defineProperty(exports, "HeaderContext", {
  enumerable: true,
  get: function get() {
    return _HeaderContext.default;
  }
});
Object.defineProperty(exports, "Epic", {
  enumerable: true,
  get: function get() {
    return _Epic.default;
  }
});
Object.defineProperty(exports, "Tabbar", {
  enumerable: true,
  get: function get() {
    return _Tabbar.default;
  }
});
Object.defineProperty(exports, "TabbarItem", {
  enumerable: true,
  get: function get() {
    return _TabbarItem.default;
  }
});
Object.defineProperty(exports, "InnerScroll", {
  enumerable: true,
  get: function get() {
    return _InnerScroll.default;
  }
});
Object.defineProperty(exports, "HorizontalScroll", {
  enumerable: true,
  get: function get() {
    return _HorizontalScroll.default;
  }
});
Object.defineProperty(exports, "FixedLayout", {
  enumerable: true,
  get: function get() {
    return _FixedLayout.default;
  }
});
Object.defineProperty(exports, "NativePopouts", {
  enumerable: true,
  get: function get() {
    return _NativePopouts.default;
  }
});
Object.defineProperty(exports, "PopoutWrapper", {
  enumerable: true,
  get: function get() {
    return _PopoutWrapper.default;
  }
});
Object.defineProperty(exports, "Alert", {
  enumerable: true,
  get: function get() {
    return _Alert.default;
  }
});
Object.defineProperty(exports, "ActionSheet", {
  enumerable: true,
  get: function get() {
    return _ActionSheet.default;
  }
});
Object.defineProperty(exports, "ActionSheetItem", {
  enumerable: true,
  get: function get() {
    return _ActionSheetItem.default;
  }
});
Object.defineProperty(exports, "ScreenSpinner", {
  enumerable: true,
  get: function get() {
    return _ScreenSpinner.default;
  }
});
Object.defineProperty(exports, "Snackbar", {
  enumerable: true,
  get: function get() {
    return _Snackbar.default;
  }
});
Object.defineProperty(exports, "ModalRoot", {
  enumerable: true,
  get: function get() {
    return _ModalRoot.default;
  }
});
Object.defineProperty(exports, "ModalPage", {
  enumerable: true,
  get: function get() {
    return _ModalPage.default;
  }
});
Object.defineProperty(exports, "ModalPageHeader", {
  enumerable: true,
  get: function get() {
    return _ModalPageHeader.default;
  }
});
Object.defineProperty(exports, "ModalCard", {
  enumerable: true,
  get: function get() {
    return _ModalCard.default;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function get() {
    return _Button.default;
  }
});
Object.defineProperty(exports, "CellButton", {
  enumerable: true,
  get: function get() {
    return _CellButton.default;
  }
});
Object.defineProperty(exports, "Header", {
  enumerable: true,
  get: function get() {
    return _Header.default;
  }
});
Object.defineProperty(exports, "Group", {
  enumerable: true,
  get: function get() {
    return _Group.default;
  }
});
Object.defineProperty(exports, "List", {
  enumerable: true,
  get: function get() {
    return _List.default;
  }
});
Object.defineProperty(exports, "Cell", {
  enumerable: true,
  get: function get() {
    return _Cell.default;
  }
});
Object.defineProperty(exports, "ListItem", {
  enumerable: true,
  get: function get() {
    return _Cell.default;
  }
});
Object.defineProperty(exports, "Footer", {
  enumerable: true,
  get: function get() {
    return _Footer.default;
  }
});
Object.defineProperty(exports, "InfoRow", {
  enumerable: true,
  get: function get() {
    return _InfoRow.default;
  }
});
Object.defineProperty(exports, "Entity", {
  enumerable: true,
  get: function get() {
    return _Entity.default;
  }
});
Object.defineProperty(exports, "Gallery", {
  enumerable: true,
  get: function get() {
    return _Gallery.default;
  }
});
Object.defineProperty(exports, "Avatar", {
  enumerable: true,
  get: function get() {
    return _Avatar.default;
  }
});
Object.defineProperty(exports, "Progress", {
  enumerable: true,
  get: function get() {
    return _Progress.default;
  }
});
Object.defineProperty(exports, "Search", {
  enumerable: true,
  get: function get() {
    return _Search.default;
  }
});
Object.defineProperty(exports, "Tabs", {
  enumerable: true,
  get: function get() {
    return _Tabs.default;
  }
});
Object.defineProperty(exports, "TabsItem", {
  enumerable: true,
  get: function get() {
    return _TabsItem.default;
  }
});
Object.defineProperty(exports, "FixedTabs", {
  enumerable: true,
  get: function get() {
    return _FixedTabs.default;
  }
});
Object.defineProperty(exports, "Spinner", {
  enumerable: true,
  get: function get() {
    return _Spinner.default;
  }
});
Object.defineProperty(exports, "PullToRefresh", {
  enumerable: true,
  get: function get() {
    return _PullToRefresh.default;
  }
});
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function get() {
    return _Link.default;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function get() {
    return _Tooltip.default;
  }
});
Object.defineProperty(exports, "Counter", {
  enumerable: true,
  get: function get() {
    return _Counter.default;
  }
});
Object.defineProperty(exports, "UsersStack", {
  enumerable: true,
  get: function get() {
    return _UsersStack.default;
  }
});
Object.defineProperty(exports, "Separator", {
  enumerable: true,
  get: function get() {
    return _Separator.default;
  }
});
Object.defineProperty(exports, "Placeholder", {
  enumerable: true,
  get: function get() {
    return _Placeholder.default;
  }
});
Object.defineProperty(exports, "FormLayout", {
  enumerable: true,
  get: function get() {
    return _FormLayout.default;
  }
});
Object.defineProperty(exports, "FormLayoutGroup", {
  enumerable: true,
  get: function get() {
    return _FormLayoutGroup.default;
  }
});
Object.defineProperty(exports, "FormStatus", {
  enumerable: true,
  get: function get() {
    return _FormStatus.default;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function get() {
    return _Switch.default;
  }
});
Object.defineProperty(exports, "File", {
  enumerable: true,
  get: function get() {
    return _File.default;
  }
});
Object.defineProperty(exports, "Input", {
  enumerable: true,
  get: function get() {
    return _Input.default;
  }
});
Object.defineProperty(exports, "Slider", {
  enumerable: true,
  get: function get() {
    return _Slider.default;
  }
});
Object.defineProperty(exports, "RangeSlider", {
  enumerable: true,
  get: function get() {
    return _RangeSlider.default;
  }
});
Object.defineProperty(exports, "Textarea", {
  enumerable: true,
  get: function get() {
    return _Textarea.default;
  }
});
Object.defineProperty(exports, "Radio", {
  enumerable: true,
  get: function get() {
    return _Radio.default;
  }
});
Object.defineProperty(exports, "Checkbox", {
  enumerable: true,
  get: function get() {
    return _Checkbox.default;
  }
});
Object.defineProperty(exports, "Select", {
  enumerable: true,
  get: function get() {
    return _Select.default;
  }
});
Object.defineProperty(exports, "SelectMimicry", {
  enumerable: true,
  get: function get() {
    return _SelectMimicry.default;
  }
});
Object.defineProperty(exports, "Div", {
  enumerable: true,
  get: function get() {
    return _Div.default;
  }
});
Object.defineProperty(exports, "Tappable", {
  enumerable: true,
  get: function get() {
    return _Tappable.default;
  }
});
Object.defineProperty(exports, "Touch", {
  enumerable: true,
  get: function get() {
    return _Touch.default;
  }
});
Object.defineProperty(exports, "ConfigProvider", {
  enumerable: true,
  get: function get() {
    return _ConfigProvider.default;
  }
});
Object.defineProperty(exports, "PanelSpinner", {
  enumerable: true,
  get: function get() {
    return _PanelSpinner.default;
  }
});
Object.defineProperty(exports, "PanelHeaderClose", {
  enumerable: true,
  get: function get() {
    return _PanelHeaderClose.default;
  }
});
Object.defineProperty(exports, "PanelHeaderBack", {
  enumerable: true,
  get: function get() {
    return _PanelHeaderBack.default;
  }
});
Object.defineProperty(exports, "PanelHeaderSubmit", {
  enumerable: true,
  get: function get() {
    return _PanelHeaderSubmit.default;
  }
});
Object.defineProperty(exports, "PanelHeaderEdit", {
  enumerable: true,
  get: function get() {
    return _PanelHeaderEdit.default;
  }
});
Object.defineProperty(exports, "PromoBanner", {
  enumerable: true,
  get: function get() {
    return _PromoBanner.default;
  }
});
Object.defineProperty(exports, "withInsets", {
  enumerable: true,
  get: function get() {
    return _withInsets.default;
  }
});
Object.defineProperty(exports, "withPlatform", {
  enumerable: true,
  get: function get() {
    return _withPlatform.default;
  }
});
Object.defineProperty(exports, "usePlatform", {
  enumerable: true,
  get: function get() {
    return _usePlatform.default;
  }
});
Object.defineProperty(exports, "classNames", {
  enumerable: true,
  get: function get() {
    return _classNames.default;
  }
});
Object.defineProperty(exports, "classnames", {
  enumerable: true,
  get: function get() {
    return _classNames.default;
  }
});
Object.defineProperty(exports, "animate", {
  enumerable: true,
  get: function get() {
    return _animate.default;
  }
});
Object.defineProperty(exports, "removeObjectKeys", {
  enumerable: true,
  get: function get() {
    return _removeObjectKeys.default;
  }
});
Object.defineProperty(exports, "SSRWrapper", {
  enumerable: true,
  get: function get() {
    return _SSR.SSRWrapper;
  }
});
Object.defineProperty(exports, "platform", {
  enumerable: true,
  get: function get() {
    return _platform.platform;
  }
});
Object.defineProperty(exports, "ANDROID", {
  enumerable: true,
  get: function get() {
    return _platform.ANDROID;
  }
});
Object.defineProperty(exports, "IOS", {
  enumerable: true,
  get: function get() {
    return _platform.IOS;
  }
});
Object.defineProperty(exports, "IS_PLATFORM_ANDROID", {
  enumerable: true,
  get: function get() {
    return _platform.IS_PLATFORM_ANDROID;
  }
});
Object.defineProperty(exports, "IS_PLATFORM_IOS", {
  enumerable: true,
  get: function get() {
    return _platform.IS_PLATFORM_IOS;
  }
});
Object.defineProperty(exports, "request", {
  enumerable: true,
  get: function get() {
    return _request.default;
  }
});
Object.defineProperty(exports, "querystring", {
  enumerable: true,
  get: function get() {
    return _querystring.default;
  }
});
Object.defineProperty(exports, "getClassName", {
  enumerable: true,
  get: function get() {
    return _getClassName.default;
  }
});
Object.defineProperty(exports, "isWebView", {
  enumerable: true,
  get: function get() {
    return _webview.isWebView;
  }
});
exports.v = exports.colors = void 0;

require("./lib/polyfills");

var _Root = _interopRequireDefault(require("./components/Root/Root"));

var _View = _interopRequireDefault(require("./components/View/View"));

var _Panel = _interopRequireDefault(require("./components/Panel/Panel"));

var _HeaderButton = _interopRequireDefault(require("./components/HeaderButton/HeaderButton"));

var _PanelHeader = _interopRequireDefault(require("./components/PanelHeader/PanelHeader"));

var _PanelHeaderContent = _interopRequireDefault(require("./components/PanelHeaderContent/PanelHeaderContent"));

var _HeaderContext = _interopRequireDefault(require("./components/HeaderContext/HeaderContext"));

var _Epic = _interopRequireDefault(require("./components/Epic/Epic"));

var _Tabbar = _interopRequireDefault(require("./components/Tabbar/Tabbar"));

var _TabbarItem = _interopRequireDefault(require("./components/TabbarItem/TabbarItem"));

var _InnerScroll = _interopRequireDefault(require("./components/InnerScroll/InnerScroll"));

var _HorizontalScroll = _interopRequireDefault(require("./components/HorizontalScroll/HorizontalScroll"));

var _FixedLayout = _interopRequireDefault(require("./components/FixedLayout/FixedLayout"));

var _NativePopouts = _interopRequireDefault(require("./components/NativePopouts/NativePopouts"));

var _PopoutWrapper = _interopRequireDefault(require("./components/PopoutWrapper/PopoutWrapper"));

var _Alert = _interopRequireDefault(require("./components/Alert/Alert"));

var _ActionSheet = _interopRequireDefault(require("./components/ActionSheet/ActionSheet"));

var _ActionSheetItem = _interopRequireDefault(require("./components/ActionSheetItem/ActionSheetItem"));

var _ScreenSpinner = _interopRequireDefault(require("./components/ScreenSpinner/ScreenSpinner"));

var _Snackbar = _interopRequireDefault(require("./components/Snackbar/Snackbar"));

var _ModalRoot = _interopRequireDefault(require("./components/ModalRoot/ModalRoot"));

var _ModalPage = _interopRequireDefault(require("./components/ModalPage/ModalPage"));

var _ModalPageHeader = _interopRequireDefault(require("./components/ModalPageHeader/ModalPageHeader"));

var _ModalCard = _interopRequireDefault(require("./components/ModalCard/ModalCard"));

var _Button = _interopRequireDefault(require("./components/Button/Button"));

var _CellButton = _interopRequireDefault(require("./components/CellButton/CellButton"));

var _Header = _interopRequireDefault(require("./components/Header/Header"));

var _Group = _interopRequireDefault(require("./components/Group/Group"));

var _List = _interopRequireDefault(require("./components/List/List"));

var _Cell = _interopRequireDefault(require("./components/Cell/Cell"));

var _Footer = _interopRequireDefault(require("./components/Footer/Footer"));

var _InfoRow = _interopRequireDefault(require("./components/InfoRow/InfoRow"));

var _Entity = _interopRequireDefault(require("./components/Entity/Entity"));

var _Gallery = _interopRequireDefault(require("./components/Gallery/Gallery"));

var _Avatar = _interopRequireDefault(require("./components/Avatar/Avatar"));

var _Progress = _interopRequireDefault(require("./components/Progress/Progress"));

var _Search = _interopRequireDefault(require("./components/Search/Search"));

var _Tabs = _interopRequireDefault(require("./components/Tabs/Tabs"));

var _TabsItem = _interopRequireDefault(require("./components/TabsItem/TabsItem"));

var _FixedTabs = _interopRequireDefault(require("./components/FixedTabs/FixedTabs"));

var _Spinner = _interopRequireDefault(require("./components/Spinner/Spinner"));

var _PullToRefresh = _interopRequireDefault(require("./components/PullToRefresh/PullToRefresh"));

var _Link = _interopRequireDefault(require("./components/Link/Link"));

var _Tooltip = _interopRequireDefault(require("./components/Tooltip/Tooltip"));

var _Counter = _interopRequireDefault(require("./components/Counter/Counter"));

var _UsersStack = _interopRequireDefault(require("./components/UsersStack/UsersStack"));

var _Separator = _interopRequireDefault(require("./components/Separator/Separator"));

var _Placeholder = _interopRequireDefault(require("./components/Placeholder/Placeholder"));

var _FormLayout = _interopRequireDefault(require("./components/FormLayout/FormLayout"));

var _FormLayoutGroup = _interopRequireDefault(require("./components/FormLayoutGroup/FormLayoutGroup"));

var _FormStatus = _interopRequireDefault(require("./components/FormStatus/FormStatus"));

var _Switch = _interopRequireDefault(require("./components/Switch/Switch"));

var _File = _interopRequireDefault(require("./components/File/File"));

var _Input = _interopRequireDefault(require("./components/Input/Input"));

var _Slider = _interopRequireDefault(require("./components/Slider/Slider"));

var _RangeSlider = _interopRequireDefault(require("./components/RangeSlider/RangeSlider"));

var _Textarea = _interopRequireDefault(require("./components/Textarea/Textarea"));

var _Radio = _interopRequireDefault(require("./components/Radio/Radio"));

var _Checkbox = _interopRequireDefault(require("./components/Checkbox/Checkbox"));

var _Select = _interopRequireDefault(require("./components/Select/Select"));

var _SelectMimicry = _interopRequireDefault(require("./components/SelectMimicry/SelectMimicry"));

var _Div = _interopRequireDefault(require("./components/Div/Div"));

var _Tappable = _interopRequireDefault(require("./components/Tappable/Tappable"));

var _Touch = _interopRequireDefault(require("./components/Touch/Touch"));

var _ConfigProvider = _interopRequireDefault(require("./components/ConfigProvider/ConfigProvider"));

var _PanelSpinner = _interopRequireDefault(require("./components/PanelSpinner/PanelSpinner"));

var _PanelHeaderClose = _interopRequireDefault(require("./components/PanelHeaderClose/PanelHeaderClose"));

var _PanelHeaderBack = _interopRequireDefault(require("./components/PanelHeaderBack/PanelHeaderBack"));

var _PanelHeaderSubmit = _interopRequireDefault(require("./components/PanelHeaderSubmit/PanelHeaderSubmit"));

var _PanelHeaderEdit = _interopRequireDefault(require("./components/PanelHeaderEdit/PanelHeaderEdit"));

var _PromoBanner = _interopRequireDefault(require("./components/PromoBanner/PromoBanner"));

var _withInsets = _interopRequireDefault(require("./hoc/withInsets"));

var _withPlatform = _interopRequireDefault(require("./hoc/withPlatform"));

var _usePlatform = _interopRequireDefault(require("./hooks/usePlatform"));

var _classNames = _interopRequireDefault(require("./lib/classNames"));

var _animate = _interopRequireDefault(require("./lib/animate"));

var _removeObjectKeys = _interopRequireDefault(require("./lib/removeObjectKeys"));

var _SSR = require("./lib/SSR");

var _platform = require("./lib/platform");

var _request = _interopRequireDefault(require("./lib/request"));

var _querystring = _interopRequireDefault(require("./lib/querystring"));

var _getClassName = _interopRequireDefault(require("./helpers/getClassName"));

var _webview = require("./lib/webview");

/**
 * Layout
 */

/**
 * Popouts
 */

/**
 * Modals
 */

/**
 * Blocks
 */

/**
 * @deprecated алиас для Cell. Будет удален в 3.0.0
 */

/**
 * Forms
 */

/**
 * Helpers
 */

/**
 * Advertisement
 */

/**
 * HOCs
 */

/**
 * Hooks
 */

/**
 * Utils
 */

/**
 * @deprecated будет удалено в 3.0.0
 */

/**
 * @deprecated будет удален в версии 3.0.0
 */

/**
 * @deprecated будет удален в версии 3.0.0
 */

/**
 * @deprecated будет удален в 3.0.0
 * @type {{}}
 */
var colors = {};
exports.colors = colors;
var v = process.env.VKUI_VERSION;
exports.v = v;
