// @vkontakte/vkui/dist/es6/components/PromoBanner/PromoBanner.js
import React from 'react';
import Icon24Dismiss from '@vkontakte/icons/dist/24/dismiss';
import Button from '../Button/Button';
import Cell from '../Cell/Cell';
import Avatar from '../Avatar/Avatar';
import classNames from '../../lib/classNames';

var PromoBanner = function PromoBanner(props) {
  return React.createElement("div", {
    className: classNames('PromoBanner', props.className)
  }, React.createElement("div", {
    className: "PromoBanner__head"
  }, props.bannerData.ageRestriction && React.createElement("span", {
    className: "PromoBanner__age"
  }, props.bannerData.ageRestriction, "+"), React.createElement("span", {
    className: "PromoBanner__label"
  }, props.bannerData.advertisingLabel || 'Advertisement'), !props.isCloseButtonHidden && React.createElement("div", {
    className: "PromoBanner__close",
    onClick: props.onClose
  }, React.createElement(Icon24Dismiss, null))), React.createElement("a", {
    href: props.bannerData.trackingLink,
    rel: "nofollow noopener noreferrer",
    target: "_blank",
    className: "PromoBanner__clickable-body"
  }, React.createElement("div", {
    className: "PromoBanner__content"
  }, React.createElement(Cell, {
    before: // @ts-ignore
    React.createElement(Avatar, {
      type: "image",
      size: 48,
      src: props.bannerData.iconLink,
      alt: props.bannerData.title
    }),
    asideContent: React.createElement(Button, {
      level: "outline"
    }, props.bannerData.ctaText),
    description: props.bannerData.domain
  }, props.bannerData.title))));
};

export default PromoBanner;
