// @vkontakte/vkui/dist/components/PromoBanner/PromoBanner.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _dismiss = _interopRequireDefault(require("@vkontakte/icons/dist/24/dismiss"));

var _Button = _interopRequireDefault(require("../Button/Button"));

var _Cell = _interopRequireDefault(require("../Cell/Cell"));

var _Avatar = _interopRequireDefault(require("../Avatar/Avatar"));

var _classNames = _interopRequireDefault(require("../../lib/classNames"));

var PromoBanner = function PromoBanner(props) {
  return _react.default.createElement("div", {
    className: (0, _classNames.default)('PromoBanner', props.className)
  }, _react.default.createElement("div", {
    className: "PromoBanner__head"
  }, props.bannerData.ageRestriction && _react.default.createElement("span", {
    className: "PromoBanner__age"
  }, props.bannerData.ageRestriction, "+"), _react.default.createElement("span", {
    className: "PromoBanner__label"
  }, props.bannerData.advertisingLabel || 'Advertisement'), !props.isCloseButtonHidden && _react.default.createElement("div", {
    className: "PromoBanner__close",
    onClick: props.onClose
  }, _react.default.createElement(_dismiss.default, null))), _react.default.createElement("a", {
    href: props.bannerData.trackingLink,
    rel: "nofollow noopener noreferrer",
    target: "_blank",
    className: "PromoBanner__clickable-body"
  }, _react.default.createElement("div", {
    className: "PromoBanner__content"
  }, _react.default.createElement(_Cell.default, {
    before: // @ts-ignore
    _react.default.createElement(_Avatar.default, {
      type: "image",
      size: 48,
      src: props.bannerData.iconLink,
      alt: props.bannerData.title
    }),
    asideContent: _react.default.createElement(_Button.default, {
      level: "outline"
    }, props.bannerData.ctaText),
    description: props.bannerData.domain
  }, props.bannerData.title))));
};

var _default = PromoBanner;
exports.default = _default;
